WordPress/src/main/java/org/wordpress/android/ui/domains/DomainRegistrationDetailsFragment.kt

```diff
@@ -399,188 +457,87 @@ class DomainRegistrationDetailsFragment : Fragment() {
+                val bundle = Bundle()
+                bundle.putParcelableArrayList(EXTRA_STATES, states)
+                fragment.arguments = bundle
+                return fragment
+            }
         }
-    }
-
-    private fun fetchStates() {
-        showStatesProgressIndicator()
-        toggleStatesInputFieldVisibility(false)
-        dispatcher.dispatch(SiteActionBuilder.newFetchDomainSupportedStatesAction(selectedCountry!!.code))
-    }
 
-    @Subscribe(threadMode = ThreadMode.MAIN)
-    fun onSupportedCountriesFetched(event: OnSupportedCountriesFetched) {
-        if (event.isError) {
-            hideFormProgressIndicator()
-            AppLog.e(
-                    T.DOMAIN_REGISTRATION,
-                    "An error occurred while fetching supported countries : " + event.error.message
-            )
-            ToastUtils.showToast(
-                    context,
-                    getString(R.string.domain_registration_error_fetching_contact_information, event.error.message)
-            )
-            return
+        override fun onCreate(savedInstanceState: Bundle?) {
+            super.onCreate(savedInstanceState)
+            states = arguments!!.getParcelableArrayList<SupportedStateResponse>(EXTRA_STATES)
+                    as ArrayList<SupportedStateResponse>
         }
-        supportedCountries = event.countries?.toCollection(ArrayList())
-        fetchContactInformation()
-    }
 
-    @Subscribe(threadMode = ThreadMode.MAIN)
-    fun onDomainContactFetched(event: OnDomainContactFetched) {
-        if (event.isError) {
-            hideFormProgressIndicator()
-            AppLog.e(
-                    T.DOMAIN_REGISTRATION,
-                    "An error occurred while fetching domain contact information : " + event.error.message
-            )
-            ToastUtils.showToast(
-                    context,
-                    getString(R.string.domain_registration_error_fetching_contact_information, event.error.message)
-            )
-        } else {
-            initialDomainContactModel = event.contactModel
-            if (initialDomainContactModel != null && !TextUtils.isEmpty(initialDomainContactModel!!.countryCode)) {
-                selectedCountry = supportedCountries?.firstOrNull { it.code == initialDomainContactModel!!.countryCode }
-                populateContactForm(initialDomainContactModel!!)
-                fetchStates()
-            } else {
-                hideFormProgressIndicator()
+        override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
+            val builder = AlertDialog.Builder(requireContext())
+            builder.setTitle(R.string.domain_registration_state_picker_dialog_title)
+            builder.setItems(states.map { it.name }.toTypedArray()) { _, which ->
+                if (targetFragment != null && targetFragment is OnStateSelectedListener) {
+                    (targetFragment as OnStateSelectedListener).onStateSelected(states[which])
```

Comment:
I'm doing something very similar right now and I think it makes sense to create the `ViewModel` with the `Activity` context and fetch it here and set the field directly. What do you think? This way you can avoid fragment-fragment communication and you don't need the listeners.

Reply:
This is super clever! I checked your implementation, and it works really well (I see tons of other uses for this too!) but it needs an adapter or something a bit more dynamic than the default implementation of the list in AlerDialog - I'm not sure how to update its content (from LiveData observer) after it is shown. Do you think it's ok to access `LiveData` value directly (without observer) in this case? This will make things a bit easier.

Reply:
I think your approach for the data is good 👍 . The only thing I was wondering about is the callback. Instead of casting the `targetFragment` and going through the fragment to the viewModel you can just inject the `ViewModel.Factory` and call directly the same ViewModel which handles the Fragment.